### PR TITLE
RDRP- 997 Replace toml.load with tomli.load

### PR DESCRIPTION
**Pull request by woodac**
Replaced toml.load with tomli.load, as it is faster and is supported by Python 3.11.

Added toml decode error exception as a co-pilot suggestion, not 100% its effective as it is not returning anything at the end of the statement.

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outputs/export_files.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import logging
 from datetime import datetime
 import toml
+import tomli
```
Comment (adriano-lopresti) [inline src/outputs/export_files.py:8]: the toml package is no longer used after switching to tomli. "import toml" can therefore be removed

File: src/staging/validation.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import os
 import toml
+import tomli
```
Comment (adriano-lopresti) [inline src/staging/validation.py:2]: the toml package is no longer used after switching to tomli. "import toml" can therefore be removed

File: src/utils/helpers.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 """Define helper functions to be used throughout the pipeline.."""
 import yaml
 import toml
+import tomli
```
Comment (adriano-lopresti) [inline src/utils/helpers.py:3]: the toml package is no longer used after switching to tomli. "import toml" can therefore be removed

File: src/outputs/export_files.py [hunk 4]
```diff
@@ -49,7 +50,7 @@ def get_schema_headers(config: dict):
 
     # Get the headers for each
     schema_headers_dict = {
-        output_name: toml.load(path) for output_name, path in schema_paths.items()
+        output_name: tomli.load(path) for output_name, path in schema_paths.items()
```
Comment (adriano-lopresti) [inline src/outputs/export_files.py]: unlike in "toml" , in "tomli" method load can't take in a string: it takes a binary (the one you would get with open(..) as f.
In order to read a string try "loads" (with an s) instead.

File: src/staging/validation.py [hunk 5]
```diff
@@ -26,17 +27,20 @@ def load_schema(file_path: str = "./config/contributors_schema.toml") -> dict:
 
     # Check if Data_Schema.toml exists
     if file_exists:
-        # Load toml data schema into dictionary if toml file exists
-        toml_dict = toml.load(file_path)
+        try:
+            # Open the file and load toml data schema into dictionary
+            with open(file_path, 'rb') as file:
+                toml_dict = tomli.load(file)
```
Comment (adriano-lopresti) [inline src/staging/validation.py:32]: this should be fine!

File: src/utils/helpers.py [hunk 6]
```diff
@@ -44,7 +45,7 @@ def user_config_reader(configfile: str = user_config_path) -> dict:
         {'title': 'TOML Example config', 'period': {'start_period':
         datetime.date(1990, 10, 10), 'end_period': datetime.date(2000, 10, 5)}}
     """
-    toml_dict = toml.load(configfile)
+    toml_dict = tomli.load(configfile)
```
Comment (adriano-lopresti) [inline src/utils/helpers.py]: unlike in "toml" , in "tomli" method load can't take in a string: it takes a binary (the one you would get with open(..) as f.
In order to read a string try "loads" (with an s) instead.

File: src/staging/validation.py [hunk 7]
```diff
@@ -26,17 +26,20 @@ def load_schema(file_path: str = "./config/contributors_schema.toml") -> dict:
...
-        # Return False if file does not exist
+        # Return None if file does not exist
         ValidationLogger.warning(
             "Validation schema does not exist! Path may be incorrect"
         )
-        return file_exists
-
-    return toml_dict
-
+        return None
```
Comment (Ryan2Y79) [inline src/staging/validation.py:42]: Forgive my ignorance, but the function previously returned `toml_dict`, but in these changes the function returns `None` in both scenarios. Is this correct?

**Review by Ryan2Y79 — COMMENTED** [review]
Just one question to address. Cheers.

File: src/staging/validation.py (hunk 7 above)
Comment (AnneONS) [reply src/staging/validation.py:42]: @Ryan2Y79 no, look at line 33, we return toml_dict there

**Review by AnneONS — APPROVED** [review]

**Pull request merged**